### PR TITLE
msg/async, v2: use epilogue for conveying CRC32 checksums/auth tag

### PR DESCRIPTION
**Pull request by rzarzynski**

### CODE REVIEW

File: src/msg/async/ProtocolV2.cc [hunk 1]
```diff
@@ -1185,7 +1131,27 @@ CtPtr ProtocolV2::read_frame_segment() {
...
     // ultimately replaced by `allocation policies`.
     rx_segments_data.emplace_back(ceph::bufferlist{});
-    return handle_read_frame_dispatch();
+
+    // XXX: for the sake of unified epilogue handling this becomes even
+    // uglier. We are doing early dispatch of Messages now. The overall
+    // idea is to:
+    //   1. parse ceph_msg_header2 which let us know tid, and thus pick
+    //      up appropriate rx_buffer (early dispatch aka pre-dispatch).
+    //   2. Read data field into selected rx_buffer.
```
Comment (rzarzynski) [inline src/msg/async/ProtocolV2.cc]: This is going to be dropped.

File: src/msg/async/frames_v2.h [hunk 2]
```diff
@@ -91,8 +114,21 @@ struct preamble_block_t {
 static_assert(sizeof(preamble_block_t) % CRYPTO_BLOCK_SIZE == 0);
 static_assert(std::is_standard_layout<preamble_block_t>::value);
 
+// V2 epilogue conveys integrity/authentication information. It's added
+// at the end of each frame holds:
+//  * CRC32 for MAX_NUM_SEGMENTS -- in plain mode,
+//  * cipher-specific data (e.g. auth tag for AES-GCM).
+union epilogue_block_t {
+  // TODO: add crc32 for plain mode
```
Comment (liewegas) [inline src/msg/async/frames_v2.h]: drop todo line?

File: src/msg/async/frames_v2.h [hunk 3]
```diff
@@ -63,6 +63,29 @@ struct segment_t {
...
 
+struct onwire_segment_t {
+  // crypto-processed segment can be expanded on-wire because of:
+  //  * padding to achieve CRYPTO_BLOCK_SIZE alignment,
+  //  * authentication tag. It's appended at the end of message.
+  //    See RxHandler::get_extra_size_at_final().
+  __le32 onwire_length;
+
+  struct ceph::msgr::v2::segment_t logical;
+} __attribute__((packed));
```
Comment (liewegas) [inline src/msg/async/frames_v2.h]: what's the purpose of separating onwire_segment_t from segment_t?

File: src/msg/async/frames_v2.h [hunk 4]
```diff
@@ -91,8 +114,21 @@ struct preamble_block_t {
...
 static_assert(std::is_standard_layout<preamble_block_t>::value);
 
+// V2 epilogue conveys integrity/authentication information. It's added
+// at the end of each frame holds:
+//  * CRC32 for MAX_NUM_SEGMENTS -- in plain mode,
+//  * cipher-specific data (e.g. auth tag for AES-GCM).
+union epilogue_block_t {
+  // TODO: add crc32 for plain mode
+  char auth_tag[CRYPTO_BLOCK_SIZE];
+  std::array<__le32, MAX_NUM_SEGMENTS> crc_values;
```
Comment (liewegas) [inline src/msg/async/frames_v2.h]: we also need a flags field in the epilogue (or footer?) for the message aborts, and that is needed for both crc and secure mode.

Comment (liewegas) [reply src/msg/async/frames_v2.h]: that is a TAG_MESSAGE specific thing (that's the only time we need to abort a frame send) so we could add a final "footer" segment that has the flags...

Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: I expect we want to have this information as close to frame's end as possible to avoid tx buffering. If it had been available at preamble crafting, we wouldn't have started transmitting at all, I guess.

Let me summarize the approaches:
* extra bit in `epilogue`. This would inflate the frame size but maybe just by one byte. The epilogue part will never be encrypted, so aligning to cipher's block size doesn't look necessary here IIUC.
* Extra segment (called e.g. *footer*) just for `Message`. I'm inclined to avoid doing that as we don't currently have room for fifth one (`Message` needs 4 segments at the moment due to `ceph_msg_header2`).

File: src/msg/async/frames_v2.h (hunk 2 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Sure. Thanks!

File: src/msg/async/frames_v2.h (hunk 3 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Not needed anymore as auth tag has been conceptually moved from *last segment* into *epilogue* -- `next_payload_len` for `Message` handling is abandoned. I think this allows round up to crypto block size (`::get_onwire_size`) just when allocating the buffer for storing physical (read: encrypted) representation of segment's data. Will clean this up.

File: src/msg/async/frames_v2.h (hunk 2 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Dropped.

File: src/msg/async/frames_v2.h (hunk 3 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Dropped.

File: src/msg/async/frames_v2.h (hunk 4 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Added a commit with proposal of the `flags` field being a part of `epilogue_block_t`.

Comment (liewegas) [reply src/msg/async/frames_v2.h]: Hmm, it seems right for the crc mode, but for secure mode, is there a way to include it in the encrypted stream without padding things out extensively?  Appended at the end like this it's neither encrypted nor checked for integrity, so a MITM can selectively drop messages

Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Just for GCM we could pick e.g. 120 bits as the length of auth tag but that way flexibility for potential cipher exchange in the future would be limited even more (moving to fixed-size *epilogue* had already constrained `sizeof(auth_tag)`, I'm afraid).

I'm not sure whether gluing the CRCs with auth tag and *late flags* is conceptually the right thing to do. If we somehow could ignore (or lower importance) of `ceph_msg_header::data_off`, following alternative becomes imaginable:
* make the *late flags* a part of `ceph_msg_header2`,
* transmit `ceph_msg_header2` **after** `front`, `middle`, and `data`.
* in crc mode use fixed-size epilogue, in secure allow cipher to determine it.

File: src/msg/async/frames_v2.h [hunk 5]
```diff
@@ -561,20 +603,70 @@ struct MessageHeaderFrame
     return this->payload;
   }
 
-  static MessageHeaderFrame Encode(const ceph_msg_header2 &msg_header,
-                                   const uint32_t front_len,
-                                   const uint32_t middle_len,
-                                   const uint32_t data_len)  {
+  static MessageHeaderFrame Encode(ceph::crypto::onwire::rxtx_t &session_stream_handlers,
```
Comment (rjfd) [inline src/msg/async/frames_v2.h]: Looks like this is no longer just a `MessageHeaderFrame`. Rename this to `MessageFrame` ?

File: src/msg/async/frames_v2.h [hunk 6]
```diff
@@ -338,25 +382,23 @@ struct SignedEncryptedFrame : public PayloadFrame<T, Args...> {
...
+      epilogue.integrity.crc[SegmentIndex::Frame::PAYLOAD] =
+          hdriter.crc32c(hdriter.get_remaining(), -1);
+      c.payload.append(reinterpret_cast<const char*>(&epilogue), sizeof(epilogue));
     }
     return c;
   }
 
   static T Decode(ceph::crypto::onwire::rxtx_t &session_stream_handlers,
                   ceph::bufferlist &payload) {
-    if (!session_stream_handlers.rx) {
```
Comment (rjfd) [inline src/msg/async/frames_v2.h]: Why did you remove this code block (13 lines)? Where does the decryption path lives now?

File: src/msg/async/frames_v2.h [hunk 7]
```diff
@@ -91,8 +104,30 @@ struct preamble_block_t {
...
+// V2 epilogue conveys integrity/authentication information. It's added
+// at the end of each frame holds:
+//  * CRC32 for MAX_NUM_SEGMENTS -- in plain mode,
+//  * cipher-specific data (e.g. auth tag for AES-GCM).
+// Additionally, it consists flags field used currently to abort message
+// processing. It can be set by transceiver even very lately without any
+// need for introducing buffering.
+struct epilogue_block_t {
+  union {
+    char auth_tag[CRYPTO_BLOCK_SIZE];
```
Comment (rjfd) [inline src/msg/async/frames_v2.h]: I spent sometime trying to find in the code where this `auth_tag` was being added to the payload bufferlist when encoding a frame, but couldn't find. Although when reading the epilogue from the wire and casting the bytes to this structure suddenly you use the `auth_tag`.
I'm guessing `auth_tag` is somekind of padding/signature bytes added by the encryption function and that's why I didn't saw it.

Having this in the epilogue feels misleading as any other person might feel a bit lost on how this encrypt/decrypt process is being done.
I would rather see a comment in the `SignedEncryptedFrame::Encode` saying that after encrypting the payload now has more `CRYPTO_BLOCK_SIZE` bytes, and in the read_frame code path see a different code path for reading the epilogue and the "remaining bytes" required for encrypted frames.

File: src/msg/async/frames_v2.h (hunk 4 above)
Comment (rjfd) [reply src/msg/async/frames_v2.h]: why can't we just add the flags byte to the payload of the last segment that contains data?

File: src/msg/async/frames_v2.h (hunk 7 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Being terribly misleading is not the biggest flaw, to be honest. Initially the length of *signature* in secure mode (*auth tag* in the concrete, GCM-based implementation we have now) was an implementation detail of `ceph::crypto::onwire`. This provided flexibility for cipher selection. Somebody could exchange AES-GCM with e.g. AES-CBC + SHA512 and `ProtocolV2` wouldn't notice that. Gluing the 3 things (CRCs, *auth tag* and *late flags*) altogether erased this useful trait.

File: src/msg/async/frames_v2.h (hunk 4 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: > why can't we just add the flags byte to the payload of the last segment that contains data?

That's also an option with its own consequences. One is blurring the distinction between data and metadata. Thus, the *last segment* would become **special**. I'm afraid about inflating allocation sizes -- if `data` is e.g. 4 KiB chunk, 4097 bytes would be allocated. Alternatively, we could go with two separated reads (4096 + 1) but this puts more complexity to the *last segment handler* and makes it even more special.

File: src/msg/async/frames_v2.h (hunk 6 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Decryption is performed by `::handle_read_frame_segment()`. I'm working on separation of *stream crypto processing* from *frame serialization*. They appear conceptually distant. The junction point I see is the epilogue crafting (+ maybe the debug information whether a particular frame is encryptable or not). Maybe `Frame::get_buffer(...)` taking a *processor* and calling and polymorphic `::fill_epilogue(buffers)` could do.

File: src/msg/async/frames_v2.h (hunk 5 above)
Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Yeah, that's true. I think that introducing segments diverged the `Frame` hierarchy/model far away from what we have on wire. I feel that `SignedEncryptedFrame` doesn't fit anymore (it's only used for control, single-segment frames; `Message{Header}Frame` bypasses it entirely). The same with single `payload` field of `Frame`. What we have now is closer to:

```cpp
template <class T, __u8 SegmentsNumV>
struct Frame {
protected:
  std::array<ceph::bufferlist, SegmentsNumV> segments;
  static_assert(std::size(segments) <= MAX_NUM_SEGMENTS);
```

Will fix the naming.

File: src/msg/async/frames_v2.h (hunk 4 above)
Comment (liewegas) [reply src/msg/async/frames_v2.h]: The problem with moving the msg header to the last segment is that we then wouldn't get teh tid before the data payload, which precludes an rx_buffers-like zero-copy implementation.

It really seems like we need an additional small footer segment...

Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Before going further, let me verify the design invariants/requirements: is avoiding 15 bytes of overhead worth the complexity inflation?

Comment (liewegas) [reply src/msg/async/frames_v2.h]: I don't understand why we can't do something simple like:

* for *crc*: add the flags to an epilogue_crc_t (that includes flags and a vector of crcs).  the is an assumed part of the frame sequenc eon the wire: preamble, segment(s), epilogue_crc.
* for *secure*: make an epilogue_secure_t (that includes just flags), followed by the cipher's trailing epilogue_secure_cipher_t.

Comment (rzarzynski) [reply src/msg/async/frames_v2.h]: Minutes from the stand-up:
* for now we can go with just moving the message header. `data_off` is used solely for enforcing alignment while `rx_buffers` are knocked out for a while.
* Bringing back 0c will need altering preamble format. We have 2 bytes reserved for flags/versioning.
* Determining secure signature size will be kept at the cipher's side.

File: src/msg/async/frames_v2.h [hunk 8]
```diff
@@ -55,7 +55,7 @@ enum class Tag : __u8 {
 struct segment_t {
   // TODO: this will be dropped with support for `allocation policies`.
   // We need them because of the rx_buffers zero-copy optimization.
-  static constexpr __le16 DEFERRED_ALLOCATION{0x0000};
+  static constexpr __le16 PAGE_SIZE_ALIGNMENT{4096};
```
Comment (rzarzynski) [inline src/msg/async/frames_v2.h]: Maybe just use `CEPH_BUFFER_ALLOC_UNIT` for the magic. I'm not sure we want introducing an abstraction over page size, especially as all our targets use 4k.

File: src/msg/async/frames_v2.h [hunk 9]
```diff
@@ -91,8 +104,30 @@ struct preamble_block_t {
 static_assert(sizeof(preamble_block_t) % CRYPTO_BLOCK_SIZE == 0);
 static_assert(std::is_standard_layout<preamble_block_t>::value);
 
+// Each Frame has an epilogue for integrity or authenticity validation.
+// For plain mode it's quite straightforward - the structure stores up
+// to MAX_NUM_SEGMENTS crc32 checksums, one per each segment.
+// For secure mode things become very different. The fundamental thing
```
Comment (rzarzynski) [inline src/msg/async/frames_v2.h]: Basically plain mode can be seen as special case of secure mode: encryptor doing `cipertext := plaintext` + very weak message authentication code (but with some extra integrity facilities). The sole difference I see at the moment is absence of the `CRYPTO_BLOCK_SIZE` rounding. This has potential for unification (rather post-nautilus).
